packages/next/client/components/navigation.ts

```diff
@@ -69,24 +72,76 @@ class ReadonlyURLSearchParams {
   }
 }
 
+/**
+ * parsedURLQueryToURLSearchParams converts a parsed url query to a url search
+ * params object.
+ *
+ * @param query parsed url query
+ * @returns url search params object
+ */
+function parsedURLQueryToURLSearchParams(
+  query: ParsedUrlQuery
+): URLSearchParams {
+  return new URLSearchParams(
+    Object.keys(query).reduce<[string, string][]>((acc, name) => {
+      const value = query[name]
+      if (Array.isArray(value)) {
+        acc.push(...value.map<[string, string]>((v) => [name, v]))
+      } else {
+        acc.push([name, value])
+      }
+
+      return acc
+    }, [])
+  )
+}
+
 /**
  * Get a read-only URLSearchParams object. For example searchParams.get('foo') would return 'bar' when ?foo=bar
  * Learn more about URLSearchParams here: https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams
  */
 export function useSearchParams() {
+  const router = useContext(RouterContext)
   const searchParams = useContext(SearchParamsContext)
+
   const readonlySearchParams = useMemo(() => {
-    return new ReadonlyURLSearchParams(searchParams)
-  }, [searchParams])
+    // To support migration from pages to app, this adds a workaround that'll
+    // support the pages router here too.
+    if (router) {
+      return new ReadonlyURLSearchParams(
+        parsedURLQueryToURLSearchParams(router.query)
+      )
+    }
+
+    if (searchParams) {
+      return new ReadonlyURLSearchParams(searchParams)
+    }
+
+    throw new Error('invariant at least one router was expected')
+  }, [router, searchParams])
+
   return readonlySearchParams
 }
 
 // TODO-APP: Move the other router context over to this one
 /**
  * Get the router methods. For example router.push('/dashboard')
  */
-export function useRouter(): import('../../shared/lib/app-router-context').AppRouterInstance {
-  return useContext(AppRouterContext)
+export function useRouter(): HybridRouter {
```

Comment:
The way I was thinking about this is that this always returns a `AppRouterInstance`, we'd just set `AppRouterContext` for the pages router too with it's own AppRouterInstance object that under the hood calls the existing router. Does that make sense? It would keep the type the same regardless and makes migrating easier.

Reply:
That makes sense! I can look into making these changes.